### PR TITLE
ENH: Use strongly typed enums for the `Algorithm` type in math morphology

### PR DESCRIPTION
**Pull request by jhlegarreta**
Use strongly typed enums for the `Algorithm` type in math morphology.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5

### CODE REVIEW

**Comment by jhlegarreta** [comment]
A few notes:
- The patch set is failing locally because the `itkMathematicalMorphologyExport.h` file is not found. IIUC the file needs not to be created manually, but not sure why it is not found in this case.
- Checking especially the changes to the `itk-module.cmake` and `CMakeLists.txt` would be appreciated.
- Not sure if having the actual `enum class` inside another class is preferable (i.e. `itk::MathematicalMorphologyEnums` in this case). I've seen other places where the `enum class` is not inside a class. I guess the former allows to have a more descriptive localization and potentially having all `enums` belonging to a class or module in a single file.
- Not sure if the `enum class` has to have its own test exercising the serialization method. Have not found any examples.

**Review by dzenanz — CHANGES_REQUESTED** [review]
As the link line has `lib\ITKLabelMap-5.2.lib ITKMathematicalMorphology.lib lib\ITKPath-5.2.lib`, https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/AntiAlias/itk-module.cmake needs to either `compile_depend` or plain old depend on `ITKMathematicalMorphology`

**Comment by dzenanz** [comment]
P.S. I didn't look at the rest of the changes, as they might change after making this build 😄

**Comment by jhlegarreta** [comment]
> As the link line has lib\ITKLabelMap-5.2.lib ITKMathematicalMorphology.lib lib\ITKPath-5.2.lib, https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/AntiAlias/itk-module.cmake needs to either compile_depend or plain old depend on ITKMathematicalMorphology

Thanks @dzenanz ! Let's see how 24b6e66 does on that front.

> P.S. I didn't look at the rest of the changes, as they might change after making this build smile

Sounds reasonable.

Cross-referencing [this comment](https://github.com/InsightSoftwareConsortium/ITK/pull/1741#issuecomment-607512513), since it looks related. Also, I realize the comment towards the end of the related section in the [migration guide](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/ITK5MigrationGuide.md).

**Comment by jhlegarreta** [comment]
Using `COMPILE_DEPENDS` similarly fails on my local Windows/VS 2015 build, so I am short of ideas about where to look.

File: Modules/Filtering/MathematicalMorphology/include/itkMathematicalMorphologyEnums.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkMathematicalMorphologyEnums_h
+#define itkMathematicalMorphologyEnums_h
+
+#include <iostream>
+#include "itkMathematicalMorphologyExport.h"
```
Comment (dzenanz) [inline Modules/Filtering/MathematicalMorphology/include/itkMathematicalMorphologyEnums.h:22]: This should be `#include "ITKMathematicalMorphologyExport.h"` (`ITK` uppercase.). But there is no compile attempt of corresponding .cxx file. No wonder that `ITKMathematicalMorphology.lib` does not exist. But I don't know why.

**Review by dzenanz — COMMENTED** [review]
I am out of ideas. @blowekamp @thewtex @hjmjohnson @N-Dekker can you spot the mistake?

File: Modules/Filtering/MathematicalMorphology/include/itkMathematicalMorphologyEnums.h (hunk 1 above)
Comment (jhlegarreta) [reply Modules/Filtering/MathematicalMorphology/include/itkMathematicalMorphologyEnums.h]: Nice catch ! Addressed in be5455a.

**Comment by thewtex** [comment]
> Modules/Filtering/MathematicalMorphology/src/CMakeLitsts.txt 

*CMakeLitsts.txt* -> *CMakeLists.txt* :bulb: @jhlegarreta

**Comment by dzenanz** [comment]
> As the link line has `lib\ITKLabelMap-5.2.lib ITKMathematicalMorphology.lib lib\ITKPath-5.2.lib`, https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/AntiAlias/itk-module.cmake needs to either `compile_depend` or plain old depend on `ITKMathematicalMorphology`

This might not be needed, now that Matt found the main mistake.

**Comment by jhlegarreta** [comment]
> CMakeLitsts.txt -> CMakeLists.txt 💡 @jhlegarreta

@thewtex  💯 👀 🦅 thanks !

> This might not be needed, now that Matt found the main mistake.

@dzenanz reverted !

**Review by dzenanz — APPROVED** [review]
Great job Jon!

**Comment by jhlegarreta** [comment]
The `ITK.macOS` failure (on test `itkJPEG2000Test06`) is unrelated to this patch set, so to me this is ready to be merged.

> Great job Jon!

Thanks for your revies @dzenanz . As always, feels great to contribute to ITK :100: .

**Comment by dzenanz** [comment]
Test failure is unrelated. Merging.

**Pull request merged**